spec/models/reports/repository_spec.rb

```diff
@@ -303,24 +303,55 @@ def refresh_views
+          # This patient registers June 2019, and has a visit with no BP in Sept 2019 and June 2020. They also become LTFU June 2020.
+          visit_with_no_bp_and_not_ltfu = create(:patient, full_name: "visit_with_no_bp_and_not_ltfu", recorded_at: "June 2nd 2019", assigned_facility: facility_1, registration_user: user)
+          create(:blood_sugar_with_encounter, patient: visit_with_no_bp_and_not_ltfu, recorded_at: "September 1st 2019", facility: facility_1, user: user)
+          create(:blood_sugar_with_encounter, patient: visit_with_no_bp_and_not_ltfu, recorded_at: "June 1st 2020", facility: facility_1, user: user)
+          # This patient registers June 2019, and has a BP taken every three months in 2020
+          visit_with_bp = create(:patient, full_name: "visit_with_bp", recorded_at: "June 1st 2019 00:00:00 UTC", assigned_facility: facility_1, registration_user: user)
+          [1, 4, 7, 10].each { |month_num|
+            create(:bp_with_encounter, :under_control, recorded_at: "2020-#{month_num}-01 00:00:00 UTC", facility: facility_1, patient: visit_with_bp, user: user)
+          }
 
           refresh_views
-          jan = Period.month(jan_2020)
-          repo = Reports::Repository.new(facility_1, periods: jan)
-          expect(repo.visited_without_bp_taken[facility_1.region.slug][jan]).to eq(1)
+          range = (jan_2019.to_period.."Jan 2021".to_period)
+          repo = Reports::Repository.new(facility_1, periods: range)
+
+          counts_without_ltfu = repo.visited_without_bp_taken[slug]
+          counts_with_ltfu = repo.visited_without_bp_taken(with_ltfu: true)[slug]
+          rates_without_ltfu = repo.visited_without_bp_taken_rates[slug]
+          rates_with_ltfu = repo.visited_without_bp_taken_rates(with_ltfu: true)[slug]
+          ("Jan 2019".to_period.."Aug 2019".to_period).each { |period|
+            expect(counts_without_ltfu[period]).to eq(0)
+            expect(counts_with_ltfu[period]).to eq(0)
+            expect(rates_without_ltfu[period]).to eq(0)
+          }
+          ("Sep 2019".to_period.."Nov 2019".to_period).each { |period|
+            expect(counts_without_ltfu[period]).to eq(1)
+            expect(rates_without_ltfu[period]).to eq(33)
+            unless v2_flag # ignore bug in v1
+              expect(counts_with_ltfu[period]).to eq(1)
+              expect(rates_with_ltfu[period]).to eq(33)
+            end
+          }
+          ("June 2020".to_period.."August 2020".to_period).each { |period|
+            expect(counts_without_ltfu[period]).to eq(0)
+            expect(rates_without_ltfu[period]).to eq(0)
+            if v2_flag # ignore bug in v1
+              expect(counts_with_ltfu[period]).to eq(1)
+              expect(rates_with_ltfu[period]).to eq(33)
+            end
+          }
+          ("Sept 2020".to_period.."Jan 2021".to_period).each { |period|
+            expect(counts_without_ltfu[period]).to eq(0)
+            expect(counts_with_ltfu[period]).to eq(0)
+          }
```

Comment:
This was the best I got to to make this specs readable while still covering a wide range of periods and data.